pkg/provider/middleware.go

```diff
@@ -0,0 +1,130 @@
+	rateLimitAverage, hasRateLimitAverage := svc.Annotations[k8s.AnnotationRateLimitAverage]
+	rateLimitBurst, hasRateLimitBurst := svc.Annotations[k8s.AnnotationRateLimitBurst]
+
+	if hasRateLimitAverage && hasRateLimitBurst {
+		average, err := strconv.Atoi(rateLimitAverage)
+		if err != nil {
+			return nil, fmt.Errorf("unable to build rate-limit middleware, %q annotaiton is invalid: %w", k8s.AnnotationRateLimitAverage, err)
+		}
+
+		burst, err := strconv.Atoi(rateLimitBurst)
+		if err != nil {
+			return nil, fmt.Errorf("unable to build rate-limit middleware, %q annotaiton is invalid: %w", k8s.AnnotationRateLimitBurst, err)
+		}
+
+		if burst <= 0 || average <= 0 {
+			return nil, errors.New("unable to build rate-limit middleware, burst and average must be greater than 0")
+		}
+
+		middleware.RateLimit = &dynamic.RateLimit{
+			Average: int64(average),
+			Burst:   int64(burst),
+		}
+	}
+
+	if middleware.CircuitBreaker == nil && middleware.Retry == nil && middleware.RateLimit == nil {
+		return nil, nil
+	}
+
+	return &middleware, nil
+}
+
+// buildWhitelistMiddlewareFromTrafficTargetDirect builds an IPWhiteList middleware which blocks every requests except
+// those originating from an authorized Pod. Authorized Pods are all the Pods listed in the ServiceTrafficTarget.Sources.
+// This middleware doesn't work if there's a proxy between the authorized Pod and this Maesh proxy.
+func buildWhitelistMiddlewareFromTrafficTargetDirect(tt *topology.ServiceTrafficTarget) *dynamic.Middleware {
+	var IPs []string
+
+	for _, source := range tt.Sources {
+		for _, pod := range source.Pods {
+			IPs = append(IPs, pod.IP)
+		}
+	}
+
+	return &dynamic.Middleware{
+		IPWhiteList: &dynamic.IPWhiteList{
+			SourceRange: IPs,
+		},
+	}
+}
+
+// buildWhitelistMiddlewareFromTrafficSplitDirect builds an IPWhiteList middleware which blocks every requests except
+// those originating from an authorized Pod. Authorized pods are the pods that can access all the leaves of the
+// TrafficSplit.
+func buildWhitelistMiddlewareFromTrafficSplitDirect(ts *topology.TrafficSplit) *dynamic.Middleware {
+	var IPs []string
+
+	for _, pod := range ts.Incoming {
+		IPs = append(IPs, pod.IP)
+	}
+
+	return &dynamic.Middleware{
+		IPWhiteList: &dynamic.IPWhiteList{
+			SourceRange: IPs,
+		},
+	}
+}
+
+// buildWhitelistMiddlewareFromTrafficTargetIndirect builds an IPWhiteList middleware like
+// buildWhitelistMiddlewareFromTrafficTargetDirect except it's intended to be used when there is at least one proxy
+// between the authorized Pod and this Maesh proxy. This middleware doesn't support nested TrafficSplit.
+func buildWhitelistMiddlewareFromTrafficTargetIndirect(tt *topology.ServiceTrafficTarget) *dynamic.Middleware {
+	whitelist := buildWhitelistMiddlewareFromTrafficTargetDirect(tt)
+	whitelist.IPWhiteList.IPStrategy = &dynamic.IPStrategy{
+		Depth: 1,
+	}
+
+	return whitelist
+}
+
+// buildWhitelistMiddlewareFromTrafficSplitIndirect builds an IPWhiteList middleware like
```

Comment:
Could you please describe it so that it can be understood without having to look up the direct version? This comment would instantly become technical debt if the referenced function changed, is deleted, etc.